### DISCUSSION TITLE
How to handle upcoming operator upgrades in ArgoCD?

### DISCUSSION THREAD

**Discussion by thikade** [argoproj/argo-cd#10028] [Q&A]
Hi,
I am running a setup where ArgoCD also installs some operators from OLM which use a manual approval policy. 
This works quite well for the initial operator install by using a PostSync hook which approves the pending installplan.  

The issue that I see is that after some time there comes a patch-release for the operator and the Subscription automatically changes state to `state: UpgradePending` and then Argo's default healthcheck for *Subscription* - which recognizes this state - sets health-check state to "Progressing", meaning my Argo App's health is now stuck progressing forever, until somebody approves the installplan. My "approval" Sync Hook is not run naturally, bc there is no sync happening on that resource.

So my question is if so has run into this and found a nice way of handling this case?

One idea would be to change the prepackaged health-check of `kind: Subscription` to allow/accept a `status.condition` of `type: InstallPlanPending` as either healthy or degraded (_for manual approvals only_), but I am unsure this might side-effect the initial install ...

**Comment by QuingKhaos** [comment]
Use a `Sync` hook instead of an `PostSync` hook, which is run in the same sync wave as the subscription. So the approval hook will always be run and approves the install plan.

**Reply by thikade** [reply]
Thanks for your answer ... I am using a Sync hook now, and this does not change the behaviour. `Subscription` stays in state `UpgradePending` and thus Argo in state `Progressing` for this application. Sync state is `synced`, and so no Sync operation is triggered and my hook not run.

**Comment by lallinger-tech** [comment]
we are facing the same problem, but we do not have a Sync or PostSync hook approving the install plan. We often deploy a operator subscription with manual approval and also some extra resources in the same ArgoCD application. When we then want to update the extra resources without updating the operator the auto sync of the application does not work as it gets stuck with the subscription. Is there an easy way to disable this behaviour and have ArgoCD only stopping for the initial install plan?

**Reply by thikade** [reply]
I would recommend putting these _extra_ resources into a separate ArgoCD app. I usually use 2 apps per operator:
- operator-xxx-install
- operator-xxx-config:  <-- extra stuff

**Comment by thikade** [comment]
I have found a solution which imho fits better when using Subscriptions with _manual_ approvals: by modifying ArgoCD's original healthcheck for Subscriptions, setting health_status to `Degraded` instead of `Progressing` in that case.  
From now on, apps containing a subscription with a pending installplan will show up as synced but degraded, which is to me kind of a notification to act (ie approve installplan), but the overall Argo's status is now `synced`.

![image](https://user-images.githubusercontent.com/8811551/192287328-a426c3a5-89e4-46d1-b9da-5889408a3668.png)

Modified health-check: https://gist.github.com/thikade/9aef2964ff564f58513c37a5f1289b39#file-argocd_healthcheck-yaml-L45-L47

**Update**: I have just realized that there is another, possibly better fitting health status available: `Suspended`. Haven't tried it yet though.

**Reply by vlskrbek** [reply]
I have the same problem. But I think the setting of the subscription to different state does not resolve the original question. 
The posthook is run only when all resources are sync and healthy. And Progressing, Degraded or Suspended status does not fulfill this condition. Even Missing or Unknown status does not fulfill this. So the only possibility is to set the Healthy state.

What can lead to a problem. When you sync application then you can think everything is fine. But in this case there is a new installplan. And it is up to you if you want to install it or not. I think in this case the Suspended state is the best one. Becuase it exactly says that the  subscription is waiting to an external action. 

Unfortunately I did not find any solution for this.

If you want to run it by posthook you have to set Health status. If you want to have information that there is a new installplan then you have to set Progressing or Suspended. But then you cannot use posthook as it will not be running.

So in this case I think the only solution is write a job, that will be running in high wave and in this job check that you can do your job by yourself :-( .

**Comment by MindTooth** [comment]
Red Hat has an article on this: https://access.redhat.com/solutions/7015069
